polymorphia-backend/src/main/java/com/agh/polymorphia_backend/service/course/CourseImportUtil.java

```
@@ -0,0 +1,86 @@
+    }
+
+    public static <T, K> List<T> getModifiedEntities(List<T> request, List<T> current, Function<T, K> keyExtractor) {
+        Map<T, Long> currentIndexMap = enumerateAsMap(current);
+        Map<T, Long> requestIndexMap = enumerateAsMap(request);
+
+        Map<K, T> currentMap = current.stream()
+                .collect(Collectors.toMap(keyExtractor, item -> item));
+
+        Map<K, Long> currentKeyIndexMap = current.stream()
+                .collect(Collectors.toMap(keyExtractor, currentIndexMap::get));
+
+        return request.stream()
+                .filter(item -> {
+                    K key = keyExtractor.apply(item);
+                    if (!currentMap.containsKey(key)) {
+                        return false;
+                    }
+                    boolean contentChanged = !item.equals(currentMap.get(key));
+                    boolean orderChanged = !requestIndexMap.get(item).equals(currentKeyIndexMap.get(key));
+                    return contentChanged || orderChanged;
+                })
+                .collect(Collectors.toList());
+    }
+
+    public static <T> Map<T, Long> enumerateAsMap(List<T> list) {
+        return IntStream.range(0, list.size())
+                .boxed()
+                .collect(Collectors.toMap(list::get, Integer::longValue));
+    }
+
+    public static <T> void handleTypeChangedObjects(
+            List<T> newObjects,
+            List<T> deletedObjects,
+            List<T> modifiedObjects,
+            List<T> currentConfig,
+            Function<T, String> keyExtractor,
+            Function<T, ?> typeExtractor
+    ) {
+        Map<String, T> currentItemsByKey = currentConfig.stream()
+                .collect(Collectors.toMap(keyExtractor, Function.identity()));
+
+        List<T> typeChangedItems = modifiedObjects.stream()
+                .filter(modified -> {
+                    T current = currentItemsByKey.get(keyExtractor.apply(modified));
+                    return current != null &&
```

`current == null` should never happen, right? If so, I'd throw a 500 here instead of silently ignoring it.